site/content/posts/2023-09-18-multiple-idps-and-identity-transformations.md

```diff
@@ -0,0 +1,129 @@
+authenticate into the clusters which trust that FederationDomain.
+
+Why would an administrator want to use multiple identity providers? Here are some examples:
+- You might like to configure multiple FederationDomains, each using a different identity provider.
+  Each FederationDomain can be used to provide authentication to a group of Kubernetes clusters.
+  For example, you could configure a FederationDomain for each division of your R&D department,
+  thus preventing developers of each division from using each other's clusters.
+- Within a single FederationDomain, you might like to use one identity provider for admin-level
+  users, while using another identity provider for developer-level users, thus allowing you to draw identities for
+  different roles from different user databases.
+- Within a single FederationDomain, you might like to allow users from two different organizations to both
+  be able to authenticate into the same clusters by configuring an identity provider for each. This could allow
+  multiple teams to collaborate on the same clusters.
+- Within a single FederationDomain, you might like to use the same external identity provider configured
+  different ways to get different types of users. For example, you could configure one OIDCIdentityProvider resource
+  to use a client ID for your human users which requires them to authenticate using multi-factor
+  authentication (configured as a setting on that client ID), while configuring another OIDCIdentityProvider
+  (with a different client ID) to allow CI/CD "bot" users to authenticate with a single factor to ease automation.
+- You might like to configure multiple LDAPIdentityProviders to each search for users or groups in a different branch
+  of your LDAP tree. Then you could use each LDAPIdentityProvider in a different FederationDomain, or you could
+  use multiple within the same FederationDomain.
+
+Configuring a FederationDomain to use one or more identity providers is as easy as listing `objectRefs`
+in the FederationDomain's spec for each OIDCIdentityProviders, LDAPIdentityProviders, or ActiveDirectoryIdentityProviders
+that you wish to use in the FederationDomain.
+
+## New: Identity transformations and policies
+
+Now that you can configure a FederationDomain to allow users from multiple identity providers to authenticate,
+what happens if one person authenticates as username "ryan" from one provider, while another person also authenticates
+as username "ryan" from a different provider? Well, Kubernetes usernames are just strings, so those two people would be
+considered the same user by Kubernetes. The same goes for group names. If conflicting usernames and group names are
+not desirable, how can you configure Pinniped to work around this? The answer is the new identity transformations feature.
+
+**Identity transformations** are Common Expression Language (CEL) expressions that may be configured on each
```

Comment:
Let's link this. 
https://github.com/google/cel-spec

Let's also reference how it has been adopted as the standard for policy in kubernetes:
https://kubernetes.io/docs/reference/using-api/cel/#:~:text=The%20Common%20Expression%20Language%20(CEL,and%20other%20constraints%20or%20conditions.

Reply:
The CEL spec is linked from the detailed docs. I don't feel like it needs to be linked again in what is intended to be a short blog post. I also don't think we need to defend our design choices in a short blog post which is only intended to announce the 0.26.0 release.

Reply:
A link won't lengthen the blog post.  Links do create value from a search point of view, so links existing on another page don't assist this page in its value.

Reply:
I don't see it so much as a defense as a signal of value.  We didn't stick with the original impl because the general kubernetes community went a different direction.  Thats a great thing to signal, and a blog post is a great time to indicate that.